src/stash/src/upgrade.rs

```diff
@@ -23,32 +33,75 @@ pub enum MigrationAction<K1, K2, V2> {
     Update(K1, (K2, V2)),
 }
 
-/// TODO(parkmycar): Once the Stash / the [`crate::Data`] trait uses [`prost`], get rid of this.
-pub trait DataProto: ::prost::Message + std::cmp::Ord + Default + Send + Sync {}
-impl<T: ::prost::Message + std::cmp::Ord + Default + Send + Sync> DataProto for T {}
-
 impl<K, V> TypedCollection<K, V>
 where
-    K: DataProto,
-    V: DataProto,
+    K: Data,
+    V: Data,
 {
     /// Provided a closure, will migrate a [`TypedCollection`] of types `K` and `V` to types `K2`
     /// and `V2`.
-    ///
-    /// TODO(parkmycar): Fill in this implementation in a future PR.
-    #[allow(clippy::unused_async)]
     pub async fn migrate_to<K2, V2>(
         &self,
-        _tx: &mut Transaction<'_>,
-        _f: impl for<'a> FnOnce(&'a BTreeMap<K, V>) -> Vec<MigrationAction<K, K2, V2>>,
+        tx: &mut Transaction<'_>,
+        f: impl for<'a> FnOnce(&'a BTreeMap<K, V>) -> Vec<MigrationAction<K, K2, V2>>,
     ) -> Result<(), StashError>
     where
-        K2: DataProto,
-        V2: DataProto,
+        K2: Data,
+        V2: Data,
     {
-        Err(StashError {
-            inner: InternalStashError::Other("unimplemented".to_string()),
-        })
+        // Create a batch that we'll write to.
+        let collection = tx.collection::<K, V>(self.name).await?;
+        let lower = tx.upper(collection.id).await?;
+        let mut batch = collection.make_batch_lower(lower)?;
+        let current = match tx.peek_one(collection).await {
+            Ok(set) => set,
+            Err(err) => match err.inner {
+                InternalStashError::PeekSinceUpper(_) => {
+                    // If the upper isn't > since, bump the upper and try again to find a sealed
+                    // entry. Do this by appending the empty batch which will advance the upper.
+                    tx.append(vec![batch]).await?;
+                    let lower = tx.upper(collection.id).await?;
+                    batch = collection.make_batch_lower(lower)?;
+                    tx.peek_one(collection).await?
+                }
+                _ => return Err(err),
```

Comment:
Yeah I'm pretty sure this is just a single retry. I'm not sure what would cause this to fail, honestly I copied it from other Stash methods, [example](https://github.com/MaterializeInc/materialize/blob/main/src/stash/src/lib.rs#L615-L622).

Reply:
In that case it's probably right? @mjibson would probably know since I'm pretty sure he wrote the linked code.

Reply:
Yes, this does a single retry. It is a pattern used in all peek fns, which need a sealed timestamp. At init time when since == upper, there's no sealed timestamp, so the peek fns forcibly advance upper by 1.